Orion/Npcs/NpcService.cs

```diff
@@ -0,0 +1,196 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Xna.Framework;
+using OTAPI.Core;
+using Orion.Framework;
+using Orion.Items;
+using Orion.Npcs.Events;
+
+namespace Orion.Npcs
+{
+	/// <summary>
+	/// Manages <see cref="INpc"/> instances.
+	/// </summary>
+	[Service("Npc Service", Author = "Nyx Studios")]
+	public class NpcService : SharedService, INpcService
+	{
+		private readonly INpc[] _npcs;
+
+		/// <inheritdoc/>
+		public int BaseNpcSpawningLimit
+		{
+			get { return Terraria.NPC.defaultMaxSpawns; }
+			set { Terraria.NPC.defaultMaxSpawns = value; }
+		}
+
+		/// <inheritdoc/>
+		public int BaseNpcSpawningRate
+		{
+			get { return Terraria.NPC.defaultSpawnRate; }
+			set { Terraria.NPC.defaultSpawnRate = value; }
+		}
+
+		/// <inheritdoc/>
+		public event EventHandler<NpcDroppedLootEventArgs> NpcDroppedLoot;
+
+		/// <inheritdoc/>
+		public event EventHandler<NpcDroppingLootEventArgs> NpcDroppingLoot;
+
+		/// <inheritdoc/>
+		public event EventHandler<NpcKilledEventArgs> NpcKilled;
+
+		/// <inheritdoc/>
+		public event EventHandler<NpcSetDefaultsEventArgs> NpcSetDefaults;
+
+		/// <inheritdoc/>
+		public event EventHandler<NpcSettingDefaultsEventArgs> NpcSettingDefaults;
+
+		/// <inheritdoc/>
+		public event EventHandler<NpcSpawnedEventArgs> NpcSpawned;
+
+		/// <inheritdoc/>
+		public event EventHandler<NpcSpawningEventArgs> NpcSpawning;
+
+		/// <inheritdoc/>
+		public event EventHandler<NpcStruckEventArgs> NpcStruck;
+
+		/// <inheritdoc/>
+		public event EventHandler<NpcTransformedEventArgs> NpcTransformed;
+
+		/// <inheritdoc/>
+		public event EventHandler<NpcTransformingEventArgs> NpcTransforming;
+
+		public NpcService(Orion orion) : base(orion)
+		{
+			_npcs = new INpc[Terraria.Main.npc.Length];
+			Hooks.Npc.PostDropLoot = InvokeNpcDroppedLoot;
+			Hooks.Npc.PreDropLoot = InvokeNpcDroppingLoot;
+			Hooks.Npc.Killed = InvokeNpcKilled;
+			Hooks.Npc.PostSetDefaultsById = InvokeNpcSetDefaults;
+			Hooks.Npc.PreSetDefaultsById = InvokeNpcSettingDefaults;
+			Hooks.Npc.Create = InvokeNpcSpawned;
+			Hooks.Npc.Spawn = InvokeNpcSpawning;
+			Hooks.Npc.Strike = InvokeNpcStruck;
+			Hooks.Npc.PostTransform = InvokeNpcTransformed;
+			Hooks.Npc.PreTransform = InvokeNpcTransforming;
+		}
+
+		/// <inheritdoc/>
+		public INpc SpawnNpc(int type, Vector2 position)
+		{
+			var terrariaNpc = new Terraria.NPC();
+			var npc = new Npc(terrariaNpc);
+			npc.SetDefaults((NpcType)type);
+			npc.Position = position;
+			return npc;
+		}
+
+		/// <inheritdoc/>
+		public IEnumerable<INpc> FindNpcs(Predicate<INpc> predicate = null)
+		{
+			var npcs = new List<INpc>();
+			for (int i = 0; i < Terraria.Main.npc.Length; i++)
+			{
+				if (_npcs[i]?.WrappedNpc != Terraria.Main.npc[i])
+				{
+					_npcs[i] = new Npc(Terraria.Main.npc[i]);
```

Comment:
Why is this assignment happening? Is this just to sync up the `_npcs` array with the one in `Terraria.Main`?